src/engine/ExecuteUpdate.cpp

```diff
@@ -0,0 +1,148 @@
+    return std::get<Id>(defaultGraph);
+  }();
+  auto transformGraph =
+      [&vocab, &localVocab, &defaultGraphIri,
+       &variableColumns](SparqlTripleSimpleWithGraph::Graph graph) {
+        return std::visit(
+            ad_utility::OverloadCallOperator{
+                [&defaultGraphIri](const std::monostate&) -> IdOrVariableIndex {
+                  return defaultGraphIri;
+                },
+                [&vocab, &localVocab](const Iri& iri) -> IdOrVariableIndex {
+                  ad_utility::triple_component::Iri i =
+                      ad_utility::triple_component::Iri::fromIriref(iri.iri());
+                  return TripleComponent(i).toValueId(vocab, localVocab);
+                },
+                [&variableColumns](const Variable& var) -> IdOrVariableIndex {
+                  return variableColumns.at(var);
+                }},
+            graph);
+      };
+  auto transformSparqlTripleSimple =
+      [&transformSparqlTripleComponent,
+       &transformGraph](SparqlTripleSimpleWithGraph triple) {
+        return std::array{transformSparqlTripleComponent(std::move(triple.s_)),
+                          transformSparqlTripleComponent(std::move(triple.p_)),
+                          transformSparqlTripleComponent(std::move(triple.o_)),
+                          transformGraph(std::move(triple.g_))};
+      };
+  return {
+      ad_utility::transform(std::move(triples), transformSparqlTripleSimple),
+      std::move(localVocab)};
+}
+
+// _____________________________________________________________________________
+std::optional<Id> ExecuteUpdate::resolveVariable(const IdTable& idTable,
+                                                 const size_t& row,
+                                                 IdOrVariableIndex idOrVar) {
+  return std::visit(
+      ad_utility::OverloadCallOperator{
+          [&idTable, &row](
+              const ColumnIndexAndTypeInfo& columnInfo) -> std::optional<Id> {
+            const auto id = idTable(row, columnInfo.columnIndex_);
+            return id.isUndefined() ? std::optional<Id>{} : id;
+          },
+          [](const Id& id) -> std::optional<Id> {
+            return id.isUndefined() ? std::optional<Id>{} : id;
+          }},
+      idOrVar);
+}
+
+// _____________________________________________________________________________
+void ExecuteUpdate::computeAndAddQuadsForResultRow(
+    std::vector<TransformedTriple>& templates, std::vector<IdTriple<>>& result,
+    const IdTable& idTable, const uint64_t row) {
+  for (const auto& [s, p, o, g] : templates) {
+    auto subject = ExecuteUpdate::resolveVariable(idTable, row, s);
+    auto predicate = ExecuteUpdate::resolveVariable(idTable, row, p);
+    auto object = ExecuteUpdate::resolveVariable(idTable, row, o);
+    auto graph = ExecuteUpdate::resolveVariable(idTable, row, g);
+
+    if (!subject.has_value() || !predicate.has_value() || !object.has_value() ||
+        !graph.has_value()) {
+      continue;
```

Comment:
There are ways to quickly make the [s, p, o, g] an array, and then use ranges for the code.
e.g.
```
auto input = std::array{&s, &o, &p, &g}; // No references in std::array...
std::array<optional<Id>, 4> optIds;
std::ranges::transform(...) // Fill the optIds
if (any in optIds unbound).
etc.

I think a good way would be to have something like `ad_utility::transform`  but for `std::array`  (something that takes an array and returns an array again.
Maybe this (very useful) helper can also be part of this PR.

Reply:
There already is a function `ad_utility::transformArray` which does what the name suggests. I had some issues while implementing your suggestion. But the problem probably only is lacking C++ knowledge on my side. Maybe we can look at that together.